src/wallet/api.rs

```diff
@@ -1255,63 +1260,97 @@ impl Wallet {
                 let witness_weight = spend_info.estimate_witness_size();
-                base_weight + witness_weight as u64
+                input_base_weight + witness_weight as u64
             })
             .sum::<u64>()
             / unspents.len() as u64;
 
-        // Note: Consider more sophisticated grouping policies in the future
-        // Group UTXOs by address
-        // TODO: Change this to HashMap<String, Vec<(ListUnspentResultEntry, UTXOSpendInfo)>>
-        let mut address_groups: Vec<(String, Vec<(ListUnspentResultEntry, UTXOSpendInfo)>)> =
-            Vec::new();
+        let target_amount = amount.to_sat();
 
-        for (utxo, spend_info) in &unspents {
+        let mut address_groups: HashMap<String, Vec<(ListUnspentResultEntry, UTXOSpendInfo)>> =
+            HashMap::new();
+        for (utxo, spend_info) in unspents {
             let address_str = utxo
                 .address
                 .as_ref()
                 .map(|addr| addr.clone().assume_checked().to_string())
                 .unwrap_or_else(|| format!("script_{}", utxo.script_pub_key));
+            address_groups
+                .entry(address_str)
+                .or_default()
+                .push((utxo, spend_info));
+        }
+
+        let (mut grouped_addresses, single_addresses): (Vec<_>, Vec<_>) = address_groups
+            .into_values()
+            .partition(|group| group.len() > 1);
+
+        // Sort by value
+        grouped_addresses
+            .sort_by_key(|group| group.iter().map(|(u, _)| u.amount.to_sat()).sum::<u64>());
+
+        // Mojo's Ultra-Compressed: Single loop with let binding
+        let (selected_utxos, selected_total, selected_weight) = {
+            let mut result_utxos = Vec::new();
+            let mut result_total = 0u64;
+            let mut result_weight = 0u64;
+
+            for group in grouped_addresses {
+                let group_total: u64 = group.iter().map(|(u, _)| u.amount.to_sat()).sum();
+                let group_weight: u64 = group
+                    .iter()
+                    .map(|(_, spend_info)| {
+                        32 + 4 + 4 + 1 + spend_info.estimate_witness_size() as u64
+                    })
+                    .sum();
 
-            match address_groups
-                .iter_mut()
-                .find(|(addr, _)| addr == &address_str)
-            {
-                Some(group) => group.1.push((utxo.clone(), spend_info.clone())),
-                None => {
-                    address_groups.push((address_str, vec![(utxo.clone(), spend_info.clone())]))
+                // Inline fee calculation (3 lines)
+                let tx_weight = 43 + result_weight + group_weight + target_weight.to_wu();
+                let estimated_fee = calculate_fee(tx_weight / 4, feerate as f32)
+                    .map_err(|e| WalletError::General(format!("Fee calculation failed: {e}")))?;
```

Comment:
This is good point to wrap the SelectionError into WalletError as a variant, then  we can simply `?` it. There are bunch of other spots here where this map is used. Should be removed everywhere.